### PR TITLE
Another implementation of case-insensitive support for subcommands and options (#9)

### PR DESCRIPTION
**Pull request by NewbieOrange**
This is an alternative PR to #1020 , follows a somewhat different approach, trying to solve #9 .

Currently it is only a draft, feedbacks are welcomed. Tests are partially added, not fully tested.

No previous tests are broken due to the new changes.

### CODE REVIEW

File: src/main/java/picocli/CommandLine.java [hunk 1]
```diff
@@ -5565,6 +5781,28 @@ protected CommandSpec commandLine(CommandLine commandLine) {
...
+             * @since 4.3 */
+            public CommandSpec caseInsensitive(boolean caseInsensitive) { caseInsensitiveCommands(caseInsensitive); caseInsensitiveOptions(caseInsensitive); return this; }
+            
+
+            /** Returns whether the commands is case-insensitive.
+             * @since 4.3 */
+            public boolean caseInsensitiveCommands() { return commands.isCaseInsensitive(); }
+            /** Sets the case-insensitivity of commands.
+             * @since 4.3 */
+            public CommandSpec caseInsensitiveCommands(boolean caseInsensitiveCommands) { commands.setCaseInsensitive(caseInsensitiveCommands);  return this; }
```
Comment (NewbieOrange) [inline src/main/java/picocli/CommandLine.java:5797]: Probably catch the IllegalStateException from LinkedCaseAwareMap and rethrow a wrapped exception here?

File: src/main/java/picocli/CommandLine.java [hunk 2]
```diff
@@ -5565,6 +5781,28 @@ protected CommandSpec commandLine(CommandLine commandLine) {
...
+
+            /** Returns whether the options is case-insensitive.
+             * @since 4.3 */
+            public boolean caseInsensitiveOptions() { return optionsByNameMap.isCaseInsensitive(); }
+            /** Sets the case-insensitivity of options.
+             * @since 4.3 */
+            public CommandSpec caseInsensitiveOptions(boolean caseInsensitiveOptions) { optionsByNameMap.setCaseInsensitive(caseInsensitiveOptions); negatedOptionsByNameMap.setCaseInsensitive(caseInsensitiveOptions); posixOptionsByKeyMap.setCaseInsensitive(caseInsensitiveOptions);  return this; }
```
Comment (NewbieOrange) [inline src/main/java/picocli/CommandLine.java:5804]: Same as above.

**Comment by remkop** [comment]
I’ll do another check tomorrow. 

Can you think of tests for any edge cases introduced by this change? Like ambiguities that didn’t exist before? I can think of cases where the combination of POSIX options now resembles a long option, or a long option with the same name as a subcommand, it may be good to have tests for those. 

Can anyone think of other tests?

**Comment by NewbieOrange** [comment]
> 
> 
> I’ll do another check tomorrow.
> 
> Can you think of tests for any edge cases introduced by this change? Like ambiguities that didn’t exist before? I can think of cases where the combination of POSIX options now resembles a long option, or a long option with the same name as a subcommand, it may be good to have tests for those.
> 
> Can anyone think of other tests?

I did some preliminary tests, yes the combination of POSIX options can resemble a long option, but this behaviour seems to be the same as the original, i.e. without case-insensitive set, it is still possible to have the aforementioned situation.

**Comment by NewbieOrange** [comment]
The new LinkedCaseAwareMap should be transparent to the rest of the code, and as long as picocli is using Map#get and Map#containsKey to check for duplications, the bahaviour should be reasonable.

There will of course be new ambiguities, like -ABCD and -a, -b, -c, -d will be ambiguous with case-insensitive set, but this behaviour should be intuitive as it is the same as -abcd and -a, -b, -c, -d without case-insensitivity.

There might be some edge cases tho, but none that I am aware of, yet.

**Comment by remkop** [comment]
I see this test is failing in Travis CI, can you take a look?

```
picocli.ModelCommandSpecTest > testCommandSpecAddSubcommand_DisallowsDuplicateSubcommandAliases FAILED
    org.junit.ComparisonFailure: expected:<...'a' for subcommand '[x]' is already used by...> but was:<...'a' for subcommand '[a]' is already used by...>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at picocli.ModelCommandSpecTest.testCommandSpecAddSubcommand_DisallowsDuplicateSubcommandAliases(ModelCommandSpecTest.java:1184)
```

**Comment by NewbieOrange** [comment]
Fixed the failing test. Also, I've updated the duplicate subcommand / option exception message to use the previous registered ones case-sensitive name, e.g. If there is already `-h` and we add option `-H`, it will now throws `Option name '-h' is used`.

**Comment by remkop** [comment]
Great, thank you! I think we are getting close.

Codecov reports that code coverage will go down if this PR is merged.

Can you add tests for the code changes that are [not or only partially covered by the tests yet](https://codecov.io/gh/remkop/picocli/pull/1021/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9waWNvY2xpL0NvbW1hbmRMaW5lLmphdmE=)?

**Comment by NewbieOrange** [comment]
Since LinkedCaseAwareMap is a private class (and probably should be), some code are not called by picocli (not public methods that can reach these lines) and they are there for potential future error-checking.

I could add some tests using reflection on the custom map, but does Codecov count that as 'coverage'?

**Comment by NewbieOrange** [comment]
For example, uncovered lines are checks for non-convertible classes, null keys, removing elements, which are not called by any other picocli code.

**Comment by NewbieOrange** [comment]
Now the coverage shall increase after this PR is merged! 😃

**Comment by NewbieOrange** [comment]
Pushing a small commit to try to trigger Codecov again for a correct report.

**Comment by remkop** [comment]
That is great! Thank you very much.
This is ready to be merged.

Do you think you will be able to add some documentation to the [parser  configuration section](https://picocli.info/#_parser_configuration) of the user manual (in the `docs/index.adoc` file)? If not, I can do it.  😄

**Comment by NewbieOrange** [comment]
I am not a native English speaker and not really familiar with asciidoc 😅.

It would probably be better if you could add those documentation.

**Comment by remkop** [comment]
okay, no worries.
Thank you for the contribution!

**Comment by jerrylususu** [comment]
> okay, no worries.
> Thank you for the contribution!

I'm glad to help on writing documentation. Is `docs/index.adoc` the only file need to change? However, I need some time to look into the changes and figure out how to document it clearly.

**Comment by remkop** [comment]
Yes index.adoc is the main file for documentation. 
I’ll also make an entry in RELEASE-NOTES.md. 

To test, run:

```
gradlew ascii
```

Then check the generated HTML  in build/docs/html5/index.html
Thanks again for helping to make picocli a better library!
